src/main/scala/com/johnsnowlabs/ml/tensorflow/TensorflowNer.scala

```diff
@@ -235,20 +250,50 @@ class TensorflowNer(val tensorflow: TensorflowWrapper,
       if (validationSplit > 0.0) {
         println(s"Quality on validation dataset (${validationSplit * 100}%), validation examples = $validLength")
         outputLog(s"Quality on validation dataset (${validationSplit * 100}%), validation examples = $validLength", uuid, enableOutputLogs, outputLogsPath)
-        measure(validDataset, extended = evaluationLogExtended, includeConfidence = includeConfidence, enableOutputLogs = enableOutputLogs, outputLogsPath = outputLogsPath, batchSize = batchSize, uuid = uuid)
+        val (_, newValMicroF1) = measure(validDataset, extended = evaluationLogExtended, enableOutputLogs = enableOutputLogs, outputLogsPath = outputLogsPath, batchSize = batchSize, uuid = uuid)
+        if (useBestModel && bestModelMetric == ModelMetrics.valMicroF1) {
+          if (newValMicroF1 >= lastValMicroF1) {
+            lastCheckPoints = saveBestModel()
+            lastValMicroF1 = newValMicroF1
+          }
+        }
       }
 
       if (test.nonEmpty) {
         println("Quality on test dataset: ")
         outputLog("Quality on test dataset: ", uuid, enableOutputLogs, outputLogsPath)
-        measure(test, extended = evaluationLogExtended, includeConfidence = includeConfidence, enableOutputLogs = enableOutputLogs, outputLogsPath = outputLogsPath, batchSize = batchSize, uuid = uuid)
+        val (_, newTestMicroF1) = measure(test, extended = evaluationLogExtended, enableOutputLogs = enableOutputLogs, outputLogsPath = outputLogsPath, batchSize = batchSize, uuid = uuid)
+        if (useBestModel && bestModelMetric == ModelMetrics.testMicroF1) {
+          if (newTestMicroF1 >= lastTestMircoF1) {
+            lastCheckPoints = saveBestModel()
+            lastTestMircoF1 = newTestMicroF1
+          }
+        }
+      }
+
+      if (useBestModel && bestModelMetric == ModelMetrics.loss) {
+        if (loss < lastLoss) {
+          lastLoss = loss
+          lastCheckPoints = saveBestModel()
+        }
+      }
+
+      if (!useBestModel) {
+        lastCheckPoints = saveBestModel()
       }
 
     }
 
     if (enableOutputLogs) {
       OutputHelper.exportLogFileToS3()
     }
+
+    lastCheckPoints
+
+  }
+
+  def saveBestModel(): Session = {
+    tensorflow.getSession()
```

Comment:
@maziyarpanahi , are you sure this is going to freeze current state, and not just return a reference to a session that can be subsequently updated?